contract/src/contract.rs

```diff
@@ -28,21 +34,79 @@ pub fn instantiate(
 ) -> Result<Response, ContractError> {
     set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
 
-    let config = Config {
+    let mut config = Config {
         owner: deps.api.addr_validate(info.sender.as_str())?,
         mint_denom: format!("{}/{}/{}", "factory", env.contract.address, MINT_SYMBOL),
+        use_cw20: msg.use_cw20.unwrap_or(false),
+        burn_cw20_addr: None,
         fee_collector_addr: deps.api.addr_validate(&msg.fee_collector_addr)?,
         burn_fee: msg.burn_fee.unwrap_or(DEFAULT_BURN_FEE),
     };
 
-    CONFIG.save(deps.storage, &config)?;
+    // If msg.use_cw20 is true, we will use the LP token technique
+    // Otherwise we will use CreateDenom, if tokenfactory is not available you should use_cw20
+    match msg.burn_cw20_addr {
+        Some(burn_cw20_addr) => {
+            config.burn_cw20_addr = Some(deps.api.addr_validate(&burn_cw20_addr)?);
+        }
+        None => {}
+    }
+    let mut messages: Vec<CosmosMsg> = vec![];
 
-    Ok(
-        Response::new().add_message(<MsgCreateDenom as Into<CosmosMsg>>::into(MsgCreateDenom {
-            sender: env.contract.address.to_string(),
-            subdenom: MINT_SYMBOL.to_string(),
-        })),
-    )
+    match msg.use_cw20.unwrap_or(false) {
+        true => {
+            let ash_token_name = format!("ASH-{}", config.mint_denom);
+            // Create the LP token using instantiate2
+            let creator = deps.api.addr_canonicalize(env.contract.address.as_str())?;
+            let code_id = msg.token_code_id.unwrap_or(0);
+            let CodeInfoResponse { checksum, .. } = deps.querier.query_wasm_code_info(code_id)?;
+            let seed = format!(
+                "{}{}{}",
+                config.mint_denom,
+                info.sender.into_string(),
+                env.block.height
+            );
+            let salt = Binary::from(seed.as_bytes());
+            // let pool_lp_address = deps.api.addr_humanize(
+            //     &instantiate2_address(&checksum, &creator, &salt)
+            //         .map_err(|e| StdError::generic_err(e.to_string()))?,
+            // )?;
+            let pool_lp_address = Addr::unchecked(
+                &instantiate2_address(&checksum, &creator, &salt)
+                    .map_err(|e| StdError::generic_err(e.to_string()))?
+                    .to_string(),
+            );
+
+            let message = CosmosMsg::Wasm(WasmMsg::Instantiate2 {
+                admin: None,
+                code_id,
+                label: ash_token_name.to_owned(),
+                msg: to_binary(&TokenInstantiateMsg {
+                    name: ash_token_name,
+                    symbol: "ASH".to_string(),
```

Comment:
Shall we have different ASH symbols/labels? It might be confusing to get ASH in one chain thinking it's the same ASH on another.

Reply:
My thinking here was there will be only 1 ASH per chain generally 
Maybe this is flawed. Trying to make a brand I guess 
Will change this to something like {}_ASH

Reply:
Yeah it is 1 ash per chain, that's correct. I am just not sure if we want to keep the name the same?